lib/integrations/azureoidc/accessgraph_sync.go

```diff
@@ -0,0 +1,237 @@
+	RoleName string
+	// SubscriptionID is the Azure subscription containing resources for sync
+	SubscriptionID string
+	// AutoConfirm skips user confirmation of the operation plan if true
+	AutoConfirm bool
+	// stdout is used to override stdout output in tests.
+	stdout io.Writer
+}
+
+func newManagedIdAction(cred *azidentity.DefaultAzureCredential, clt AccessGraphAzureConfigureClient, subId string, managedId string, roleName string) (*provisioning.Action, error) {
+	customRole := "CustomRole"
+	scope := "/subscriptions/" + subId
+	runnerFn := func(ctx context.Context) error {
+		// Create the role
+		roleDefinition := armauthorization.RoleDefinition{
+			Name: &roleName,
+			Properties: &armauthorization.RoleDefinitionProperties{
+				RoleName: &roleName,
+				RoleType: &customRole,
+				Permissions: []*armauthorization.Permission{
+					{
+						Actions: libslices.ToPointers([]string{
+							"Microsoft.Compute/virtualMachines/read",
+							"Microsoft.Compute/virtualMachineScaleSets/virtualMachines/read",
+							"Microsoft.Authorization/roleDefinitions/read",
+							"Microsoft.Authorization/roleAssignments/read",
+						}),
+					},
+				},
+				AssignableScopes: []*string{&scope}, // Scope must be provided
+			},
+		}
+		roleId, err := clt.CreateRoleDefinition(ctx, scope, roleDefinition)
+		if err != nil {
+			return trace.BadParameter("failed to create custom role: %v", err)
+		}
+
+		// Assign the new role to the managed identity
+		roleAssignParams := armauthorization.RoleAssignmentCreateParameters{
+			Properties: &armauthorization.RoleAssignmentProperties{
+				PrincipalID:      &managedId,
+				RoleDefinitionID: &roleId,
+			},
+		}
+		if err = clt.CreateRoleAssignment(ctx, scope, roleAssignParams); err != nil {
+			return trace.BadParameter(
+				"failed to assign role %s to principal %s: %v", roleName, managedId, err)
+		}
+
+		// Assign the Graph API permissions to the managed identity
+		graphPrincipal, err := clt.GetServicePrincipalByAppId(ctx, graphAppId)
+		if err != nil {
+			return trace.BadParameter("failed to get the graph API service principal: %v", err)
+		}
+		for _, appRole := range graphPrincipal.AppRoles {
+			if _, ok := requiredGraphRoleNames[*appRole.Value]; ok {
+				roleAssignment := msgraph.AppRoleAssignment{
+					AppRoleID:   appRole.ID,
+					PrincipalID: &managedId,
+					ResourceID:  graphPrincipal.ID,
+				}
+				if err = clt.GrantAppRoleToServicePrincipal(ctx, roleAssignment); err != nil {
+					return trace.BadParameter("failed to assign graph API role to %s: %v", managedId, err)
+				}
+			}
+		}
+		return nil
+	}
+	cfg := provisioning.ActionConfig{
+		Name:    "NewSyncManagedId",
+		Summary: "Creates a new Azure role and attaches it to a managed identity for the Discovery service",
+		Details: strings.Join([]string{
+			"The Discovery service needs to run as a credentialed Azure managed identity. This managed identity ",
```

Comment:
I think it should be Discovery Service with a capital `S`. Trying to be consistent with docs.